### ISSUE TITLE
VoiceOver: Cannot interact with toots

### ISSUE THREAD

**Issue by MarcoZehe** [dNitza/mercury-issues#6]
Transferring this from the [discussion in the fediverse](https://toot.cafe/@marcozehe/104525120476765778) over here so it doesn't get lost.

Right now, a VoiceOver user cannot interact with a toot at all, other than passing through the gesture (double-tap and hold), and once the three ascending tones are heard, swipe left (short or long), or right (short or long), in the hopes to find the correct short or long. Also, this does not give access to tapable items such as links, mentioned users, hashtags, etc. Also, even in details view, the stats to a toot aren't being exposed.

As @rkingett suggested in the thread and other discussions, custom actions should be implemented to deal with this. The default action, meaning merely focusing and then double-tapping the toot in the timeline, should, IMO, do either of two things:

1. If the toot text is not revealed due to a content warning, reveal it and update the AccessibilityLabel so VoiceOver users can immediately read it. Also, a custom action should be present in this case to show the toot details. Another double-tap should, I think, hide the text again.
2. If no CW is present, the default action should open the toot details. In this case, no extra action to show the toot details is needed.

In both cases, additional actions should be present for replying, boosting, favoriting, and any tapable items inside, including, but not limited to:

- The author profile.
- Profiles of mentioned users.
- Hashtags.
- Links.
- Dealing with audio/video playback if that is being offered in the main timelines. If this is only possible from the toot details, discard this point.
- Any other items that are tapable.

In the toot details, everything should be reachable via swipes, including possibly the client if it is shown, the stats, etc. If replies are shown, these should again be treated like in the default timeline views, with options to open their details of course. That is, if Mercury offers this multi-level toot detail viewing.

**Comment by rkingett** [comment]
Agree with everything! I also think that double tapping on an expanded CW should hide the text again. This would make scrolling a lot quicker if you want to, say, tuck the toot away.

**Comment by dNitza** [comment]
Hey @MarcoZehe and @rkingett 

I have a TestFlight build with a bunch of VoiceOver specific changes ready for testing — https://testflight.apple.com/join/LTklV3Q2

The overview of this change is that I have implemented almost all of the things we spoke about, with a few more left to go and I've also tried to address the other VoiceOver related issues raised in issues #5 and #8  

More details and a list of known issues below :)

---

Custom actions implemented for statuses without spoiler text:

- Show status details (which shows the replies)
- Favourite
- Boost
- Reply
- Bookmark
- View author profile
- Play audio if there is any (and you can action the status again or select the play audio action again to pause)
- Play video if there is one
- Mentions
- Links
- Hashtags

For statuses that have spoiler text, the list is pretty much the same, except actioning the status will adjust the label of the status to be just the hidden text so it is read out immediately. Double tapping again will return the label to its previous state as @rkingett suggested.

I've also adjusted the text that is read back on a status to: 
- Move the posted at time to the end
- Include the number of boosts / likes / replies on a toot if there are any
- Be more clear about attachments that don't have a description — I have been planning to lean on some of Apple's machine learning stuff to try and provide fallback descriptions for these in a future release.

---

### Current known issues
- Currently actioning a status without a content warning does not show the detail screen, you still need to go to the first custom action to do this.
- Returning to the main timeline from the detail screen doesn't yet return focus to the selected status.
- I still need to handle load gap cells in the timeline.

---

### Things I am still exploring
- When viewing a status detail page of a status that has spoiler text, all replies will be behind a content warning. Is it safe to assume that one wants to see the hidden content and therefore we can forego the need to explicitly action each status to read the hidden text?
- Navigating to a status detail page will focus the selected status in context of its replies. This means there could be some statuses above and below the selected status. I am looking at ways to best communicate this.

**Comment by rkingett** [comment]
I haven't tried the test flight yet, but I think it is safe to assume that one wants to see the hidden content in threads and therefore we can forego the need to explicitly action each status to read the hidden text.

But I'm also thinking about in home timelines. I still may want those behind a CW because I am not interested in reading that thread right now.

As for the, out of context, replies. Maybe just adding, reply, or, continued, will do. Continued seems best. Not sure where to add that announcement though.

**Comment by MarcoZehe** [comment]
Wow, this is exciting! However I run into an immediate crash when choosing any timeline. I have submitted about 5 crash reports with e-mail address and comments about what I did and tried.

**Comment by dNitza** [comment]
@MarcoZehe d'oh! What an anticlimax! I'll investigate this and hopefully have a fix out as soon as I can :)

**Comment by rkingett** [comment]
Like I said in my email, I am not having the crashing problem, but I hope my device info helps resolve the crashing bug. I just have suggestions on the router order.

I'd actually put the, show details, action more towards the bottom, and then have the toot actions, reply, boost, favorite, at the top. Then, open link, if there is a link. Then, play audio/video, ten, view hashtags, then show toot details.

When I get notifications, I'll give more feedback, but I think VO should read all those Toots, notifications, mentions, with just one swipe, so I'm not swiping through paragraph breaks and usernames and otherwise in notifications.

I hope that helps!

**Comment by dNitza** [comment]
Hi @MarcoZehe — I've addressed the crash so you should be good to go once the next build hits TestFlight 👍 

The most recent build also addresses the known issues from the previous build so that means that the following are now true:

- Actioning (double tapping) a status will show the detail screen if it doesn't have a content warning.
- Returning to the main timeline from a detail screen now correctly refocuses the previously focused status.
- The cell used to load the gap in a timeline now has a better description and its own actions.

@rkingett That is super helpful feedback! I've re-arranged the actions to follow your suggestion. I'm open to adding a setting to adjust this in the future if it's something that would be useful.

As for the notifications, I will be tackling those next, they're just in a different format to the toots :) Same goes for search.

**Comment by MarcoZehe** [comment]
I am happy to report that the crash is fixed, and this app works really beautifully now! Next up, testing if I can actually post an image. Like, select one from the photo library. Many apps implement their custom photo picker and fall into accessibility traps, instead of using Apple's provided UI. ;-) I will file separate issues if I see anything.

**Comment by dNitza** [comment]
@MarcoZehe good to hear :) I have used Apple's provided UI for image picking, it is so darn good. I think there are a few variations of it though, so open to suggestions again there if I have used one of the variations that doesn't work as well. In order to add an image, there is a button to take a photo and button to select from your gallery after the text area for composing a toot. There are also a few other actions after that like setting the visibility, content warning, custom emoji and scheduling a status.

**Comment by rkingett** [comment]
This is a very small thing, but after I boost a Toot, and then go back and review my home timeline later, I forgot which Toots I boosted, and there is currently no indication that a Toot has been boosted or not.

Could either of the below work as a solution?

1. Make the boost action change to, unboost, if a Toot has been previously boosted?
2. Put a small message at the start of the item, before the name, that says, "boosted,"

I like 2 the best because I can just flick through timelines quicker that way.

**Comment by MarcoZehe** [comment]
To be clear: You would only want that "Boosted" announcement at the start of toots you boosted, correct?

**Comment by rkingett** [comment]
Correct. I'm not sure how that would be a variable, though? I'm sure I just unboosted some Toots because I forgot I boosted them. But anyway, that is correct. Pinafore does this, but the "BOOSTED" comes at the end. I'd like it at the beginning.

**Comment by dNitza** [comment]
Hey @MarcoZehe and @rkingett 

Hope you're both doing well.

I have just submitted another beta which has the following changes:

- The word "boosted" is spoken at the start of a status. There is a new option allowing this to be spoken at the end instead.
- A new accessibility options section that lets you set the order of the custom actions.
- Added voting on polls as a custom action and improved the read out of poll options. There's still some work to do on this one to improve the readout of polls that have finished and you have voted on.
- The notifications section should be easier to read now
- There was something weird going on with hashtag selection which has now also been fixed.

This release also addresses a handful of other open tickets, but I will comment on the relevant issue for those.

Cheers,
Dan.

**Comment by dNitza** [comment]
All of this is in the AppStore version of the app — thanks for all of your help 🥳
